text/0000-network-transform-buffered-interpolation.md

```diff
@@ -0,0 +1,408 @@
+# Network Transform Buffered Interpolation
+[feature]: #feature
+
+- Start Date: `2021-07-19`
+- RFC PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- SDK PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/pull/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes upgraded interpolation options from the old NetworkTransform's interpolation that hides jitter and is more flexible for users.
+
+
+# Motivation
+[motivation]: #motivation
+
+
+NetworkTransform needs a way to help users smooth movements that can resist to different network conditions and allows users to extend it if needed.
+The previous implementation didn't offer much flexibility on what interpolation algorithm to use. If I wanted to create my custom interpolation that prevents players from going through walls instead of interpolating linearly, I would need to reimplement at least part of my own NetworkTransform. If I wanted my object to change direction instantaneously I could use a different interpolation implementation than if I wanted to have wobble and inertia.
+
+It also didn't take into account network jitter, which becomes a visible issue on overloaded or unstable networks like you can find with mobile platforms 
+
+The following video shows the MLAPI-0.1.0 NetworkTransform under 5% packet loss and 60-100ms jitter.
+
+https://user-images.githubusercontent.com/71790295/126506051-ebfb1974-e4fc-44ca-9923-2be450c722b6.mp4
+
+![](0000-network-transform-buffered-interpolation/jitter.jpg)
+
+This RFC suggests an extensible interface for users to use and create different interpolation algorithms. It also suggests a default buffered interpolator that'll resist to jitter. 
+After this, users should be able to select the interpolation algorithm that best fits the GameObject they want to sync.
+
+In addition, this RFC suggests removing the 
+```C#
+        public AnimationCurve DistanceSendrate = AnimationCurve.Constant(0, 500, 20);
+```
+configuration (that was used to set interpolation times) to use instead a swappable interpolator.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+NetworkTransform now has this new configuration:
+- `Interpolator`: Allows users to select the interpolation algorithm they want to use. This interpolator will interpolate "ghost" side state changes. Note it'll still be up to users to interpolate their authoritative objects however they want. If for example Client 1 is authoritative over a player object, this RFC will add interpolation to the server object and the other clients' objects, not Client 1's object. It'll still be up to users to interpolate that authoritative object (using RigidBody's interpolation for example or anything else).
```

Comment:
This does not sound like great UX to me. Since the idea is that the position changes to the NetworkTransform happen during a fixed step (be it network tick or FixedUpdate doesn't matter) I think we should consider providing a way to auto interpolate the authoritative objects as well by using the same interpolation as a RigidBody.

Reply:
you would still need different interpolation algorithm for each. Buffered interpolation adds artificial latency. if I have 100ms of buffer, that'd mean the host would have 100ms of latency on its authoritative object for no reason.
There's not really any reason for NetworkTransform to touch an authoritative object. NetworkTransform is to ghost game owned objects.
The lerp a user would have needed to add in a singleplayer game would still be needed here. They'd just add NetworkTransform to "just make it network"